### PR TITLE
backupccl: export relations with null values to parquet

### PR DESCRIPTION
**Pull request by msbutler**
Previously, EXPORT PARQUET could only export relations without null values. Now
EXPORT PARQUET can handle relations with null values for column types
currently supported by EXPORT PARQUET.

Release note (sql change): users can now export relations with null values to
parquet files.

Informs #67710

### CODE REVIEW

File: pkg/ccl/importccl/exportcsv_test.go [hunk 1]
```diff
@@ -304,6 +304,13 @@ func validateParquetFile(t *testing.T, file string, truthRows [][]interface{}) e
 
 		t.Logf("\n Record %v:", count)
 		for i := 0; i < len(cols); i++ {
+			if truthRows[count][i] == nil {
+				// if we expect a null value, the row created by the parquet reader will not have the
+				// associated column
```
Comment (stevendanna) [inline pkg/ccl/importccl/exportcsv_test.go:309]: [nit] I think most parts of the code base try to keep comments punctuated where possible:
```suggestion
				// If we expect a null value, the row created by the parquet reader will not have the
				// associated column.
```

File: pkg/ccl/importccl/exportcsv_test.go [hunk 2]
```diff
@@ -360,6 +367,24 @@ func TestBasicParquetTypes(t *testing.T) {
 
 	err = validateParquetFile(t, paths[0], truth)
 	require.NoError(t, err)
+
+	// Test EXPORT PARQUET WITH NULL VALUES
+	sqlDB.Exec(t, `INSERT INTO foo VALUES (4, 'Alex', 3, 14.3, NULL), (5, 'Bobby', 2, NULL,false),
+(6, NULL, NULL, NULL, NULL)`)
+
+	sqlDB.Exec(t, `EXPORT INTO PARQUET 'nodelocal://0/null_parquet' FROM SELECT *
+		FROM foo ORDER BY x ASC LIMIT 2`)
```
Comment (stevendanna) [inline pkg/ccl/importccl/exportcsv_test.go:376]: Are we getting much out of including the `ORDER BY ...` clauses here?

If we add another case to this test in the next PR, let's convert it to a table-driven test.

**Review by stevendanna — COMMENTED** [review]
Seems good overall. Left a few questions but nothing major.

File: pkg/ccl/importccl/exportparquet.go [hunk 3]
```diff
@@ -159,7 +159,7 @@ func newParquetColumn(typ *types.T, name string) (parquetColumn, error) {
 			See this blog post for more on parquet type specification:
 			https://blog.twitter.com/engineering/en_us/a/2013/dremel-made-simple-with-parquet
 	*/
-	col.definition.SchemaElement.RepetitionType = parquet.FieldRepetitionTypePtr(parquet.FieldRepetitionType_REQUIRED)
+	col.definition.SchemaElement.RepetitionType = parquet.FieldRepetitionTypePtr(parquet.FieldRepetitionType_OPTIONAL)
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/importccl/exportparquet.go:162]: Interesting point. The table's schema provides info on which columns are non-nullable. To get that info to the parquet exporter, I'd have to add more info to the `ParquetWriterSpec`. In a separate branch, I'm trying to figure out how to pass column names via the `ParquetWriterSpec`, so I think I'll figure out how to pass info on non-nullable columns in that branch of work. For now, I think it's fine to set all columns to optional.

File: pkg/ccl/importccl/exportcsv_test.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/importccl/exportcsv_test.go:309]: Done.

File: pkg/ccl/importccl/exportcsv_test.go (hunk 2 above)
Comment (msbutler) [reply pkg/ccl/importccl/exportcsv_test.go:376]: Could you clarify what you mean by "table-driven test?"

Comment (stevendanna) [reply pkg/ccl/importccl/exportcsv_test.go:376]: Happy to. "Table driven tests" is the name the go community has given to structuring your tests such that you can iterate over a slice of structs and share test code between test cases. For example: https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/importccl/exportcsv_test.go#L391 and https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/importccl/import_stmt_test.go#L126

Comment (msbutler) [reply pkg/ccl/importccl/exportcsv_test.go]: done.

Comment (msbutler) [reply pkg/ccl/importccl/exportcsv_test.go]: I like the order by clauses, as I can explicitly test that the parquet exporter is ordering rows correctly

**Review by stevendanna — APPROVED** [review]
This seems like a reasonable step.  If you wanted you could move the setup for your tests into your test struct as well so each case was self-contained. But, probably best to just see what you'll need as you add more tests.

**Comment by msbutler** [comment]
bors r=stevendanna

**Pull request merged**